### DISCUSSION TITLE
Calling a Java static method against a Java class (`java_class`)

### DISCUSSION THREAD

**Discussion by dmikurube** [jruby/jruby#6634] [Ideas]
From a recommendation by @headius, let me share my use-case of a "Java class" in JRuby.
https://twitter.com/headius/status/1375130217518403586

I wanted to call a static method of a Java class from JRuby in a simple manner. Although I can call it with a JRuby wrapper like `Java::java.lang.String.join(":", "a", "b")` easily, it is not very easy to call it with a Java class. For example, `Java::java.lang.String.java_class.join(":", "a", "b")` does not work.

I know that `Java::java.lang.String.java_class.join(":", "a", "b")` was an arbitrary case. A simpified example of my actual use-case was a Java class loaded by my own class loader.

```
MY_CLASS_LOADER = new URLClassLoader(["/path/to/some.jar"], JRuby.runtime.getJRubyClassLoader())
JAVA_CLASS = MY_CLASS_LOADER.loadClass("com.example.SomeClass")
JAVA_CLASS.someStaticMethod  # => Error!
```

Instead, I had to insert `ruby_class` for the loaded Java class:

```
MY_CLASS_LOADER = new URLClassLoader(["/path/to/some.jar"], JRuby.runtime.getJRubyClassLoader())
JAVA_CLASS = MY_CLASS_LOADER.loadClass("com.example.SomeClass")
JAVA_CLASS.ruby_class.someStaticMethod  # => Success
```

The `ruby_class` was actually a sufficient workaround for me. But, let me share this case to help the JRuby team improving the `java_class` object in 9.3. Hope it helps!

**Comment by headius** [comment]
There is some unification of these types by @kares in #6513, due to ship in 9.3 "any day now". That PR unifies the JavaClass API and the java.lang.reflect.Class API by decorating the latter with the former's methods. However the result is still a reflective "mirror" object and is not a true metaclass.

@dmikurube discovered `JavaClass#ruby_class` (which I had forgotten exists) that gets you from a JavaClass to the proxy metaclass we create, which was sufficient to get at the static methods. I asked them to open this discussion so we could discuss whether any additional improvements could be made to further unify the java.lang.Class API with the proxy API.

I recall other folks asking why we don't have JavaClass/java.lang.Class methods on the proxy metaclasses, or why the static methods of a class are not bound on the JavaClass/java.lang.Class object. The only answers I could come up with were "we just haven't done it" or "the static methods might collide with an existing method" but I don't recall having had a discussion about it.

**Reply by kares** [reply]
Can confirm this is still "an issue", even after the recent unification.
JRuby's assumption is for the static methods to be called like: `Java::java.lang.String.join(":", "a", "b")` (on the proxy).

The `java.lang.String.java_class` returns a `java.lang.Class` wrapper.
Believe this aligns with Java where you `java.lang.String.join("a");` but can not `java.lang.String.class.join`.
I am not against extending the Java class wrappers but I am afraid it would introduce more issues than it solves.
